File: packages/typescript/sdk/core/__tests__/validate-params.test.ts
import { describe, expect, it } from "bun:test";
import { validateParams } from "../src/validate-params";

describe("validateParams", () => {
  it("accepts empty params when schema has no required fields", () => {
    expect(validateParams({ type: "object" }, {})).toBeNull();
    expect(validateParams(undefined, { anything: 1 })).toBeNull();
  });

  it("rejects missing required keys", () => {
    const err = validateParams({ type: "object", required: ["body"] }, {});
    expect(err).toContain("body");
    expect(err).toContain("required");
  });

  it("rejects wrong property type", () => {
    const err = validateParams(
      {
        type: "object",
        properties: { count: { type: "integer" } },
      },
      { count: "not an int" },
    );
    expect(err).toContain("count");
    expect(err).toContain("integer");
  });

  it("rejects enum mismatch", () => {
    const err = validateParams(
      {
        type: "object",
        properties: { status: { type: "string", enum: ["open", "closed"] } },
      },
      { status: "other" },
    );
    expect(err).toContain("status");
  });

  it("accepts enum object member regardless of key order", () => {
    // Python/Go/Rust compare enum members structurally; JS historically used
    // JSON.stringify which is order-sensitive. Make sure we match.
    const err = validateParams(
      {
        type: "object",
        properties: {
          target: {
            type: "object",
            enum: [{ x: 1, y: 2 }],
          },
        },
      },
      { target: { y: 2, x: 1 } },
    );
    expect(err).toBeNull();
  });

  it("validates nested arrays", () => {
    const ok = validateParams(
      {
        type: "object",
        properties: { tags: { type: "array", items: { type: "string" } } },
      },
      { tags: ["a", "b"] },
    );
    expect(ok).toBeNull();

    const err = validateParams(
      {
        type: "object",
        properties: { tags: { type: "array", items: { type: "string" } } },
      },
      { tags: ["a", 2] },
    );
    expect(err).toContain("tags[1]");
  });
});

File: packages/go/slop-ai/validate_params_test.go
package slop

import "testing"

func TestValidateParams(t *testing.T) {
	// nil schema -> ok
	if got := ValidateParams(nil, map[string]any{"x": 1}); got != "" {
		t.Fatalf("nil schema should be ok, got %q", got)
	}

	// required
	got := ValidateParams(map[string]any{"type": "object", "required": []any{"body"}}, map[string]any{})
	if got == "" {
		t.Fatalf("expected required error")
	}

	// wrong type
	schema := map[string]any{
		"type":       "object",
		"properties": map[string]any{"count": map[string]any{"type": "integer"}},
	}
	got = ValidateParams(schema, map[string]any{"count": "nope"})
	if got == "" {
		t.Fatalf("expected type error")
	}

	// array items
	schema = map[string]any{
		"type":       "object",
		"properties": map[string]any{"tags": map[string]any{"type": "array", "items": map[string]any{"type": "string"}}},
	}
	if got := ValidateParams(schema, map[string]any{"tags": []any{"a", "b"}}); got != "" {
		t.Fatalf("expected ok, got %q", got)
	}
	if got := ValidateParams(schema, map[string]any{"tags": []any{"a", 2}}); got == "" {
		t.Fatalf("expected failure on tags[1]")
	}
}
